Rewrite Sabre interfaces to be Rust native

jakelishman opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

From a pure-Rust interface perspective, there are several things going on here, ticks indicate whether they're done or not:

- [x] Move all descriptions of the routing target information and construction of them into Rust (the new `RoutingTarget`)
- [x] Move all construction of the Sabre interaction DAG to Rust (the new `SabreDAG`)
- [x] Move `SabreSwap` DAG rebuild to Rust (the new `RoutingResult` and `RoutingResult::rebuild`)
- [x] Rework Python and Rust entry points `sabre_routing`/`swap_map` to use the new structures and returns.
- [x] Rewrite Python `SabreSwap` to use the new Rust components.
- [x] Move `SabreLayout` DAG rebuild to Rust (`RoutingResult::rebuild_onto` is most of the rebuild, but it's not hooked up).
- [x] Rewrite Python-space `SabreLayout` to use the new Rust components.
- [x] Properly honour `SabreLayout`'s `skip_routing` option (rather than routing and then throwing it away again)
- [x] Handle disjoint coupling maps in `SabreLayout` using #14177.
- [x] Fix `StarPrerouting` (obsoleted by #14381).
- [x] Rebase into something coherent.
- [x] Sort out last `todo!()` stubs that apparently have no test coverage.


### Details and comments

Close #12279 
Close #12280

The main bulk of this PR is preceded by several commits to parts of `crates/circuit` that I've separated out.  Any/all of these could become separate PRs if desired; all the commits should be standalone (though I didn't test individual ones).  See the commit messages for details.

The rest of this section is about commit 0549a6dd0be45f92eeddb8c8d28f8118491d1241, which is all actual Sabre stuff (it's just the commit message).

This is a major rewrite of all the Sabre interfaces, though the core routing logic is unchanged.  Note that the test suite is unaltered.

At least as far as the test suite covers, this change is 100% RNG compatible with its parent, and in several places, explicit code is added to achieve this.  In utility-scale testing (in particular, one ASV test), several tens of thousands of iterations into one particular routing problem, there is a divergence, which I suspect to be floating-point differences given that everything else is the same.

### Changes to data structures

#### Changed `SabreDAG`

`SabreDAG` is now completely Rust-internal and not even exposed to Python.  It's constructed directly from a `DAGCircuit`.  When control-flow blocks are encountered, the temporary `DAGCircuit` backing them is stored within the `SabreDAG` node itself, so there's no need for all the `block_results` side-car handling.

The nodes in the graph explicitly store how the node affects routing: they can be `TwoQ` (the actual routing target), `Synchronize` (a non-routing object that has only data dependencies) or `ControlFlow`.  This simplifies the internal handling within the router, since now it doesn't need to re-infer it.  It also lets us decay control-flow operations to simple `Synchronize` points when we're in layout mode, similar to how clearing the `block_nodes` field could implicitly do before.

A follow-up commit (which necessarily changes the RNG compatibility) makes stronger use of this new `SabreDAG` form to compress the interaction graph during construction, reducing the amount of work done by routing trials when advancing the state after a gate is routed.

#### New `RoutingProblem`

A minor record-like struct that bundles up several related arguments.

#### New `RoutingTarget`

The old Python-space code had a 3-tuple
```rust
(NeighborTable, Graph<(), ()>, ArrayView2<f64>)
```
to represent the hardware chip the routing was taking place on.  The first two elements stem from difficulties transferring the Rustworkx Python-wrapped graph backing `CouplingMap` down to Rust space, and from trying to faithfully port the existing implementation to Rust.

The `NeighborTable` and the `Graph<(), ()>` functionally serve the same role; they both relate a node to its neighbors, with one presenting a list-like interface for efficient iteration, and one presenting a graph-like interface so graph algorithms like `dijsktra`, `token_swapper` and `adjacency_matrix` could be called on it.

Now in pure Rust space, both are replaced by a single `Neighbors` structure, which supplies both interfaces.  It implements all the relevant `petgraph::visit` traits, so it can be directly passed to `petgraph` and `rustworkx_core` graph algorithms.  Its data format is a flat CSR-like sorted list of neighbors. It's logically the same as `NeighborTable`, except all the `SmallVec`s are flattened into a single contiguous `Vec` with partition points.

The distance matrix remains the same; we need super fast lookups of scores because it's in the innermost loop of the swap-mapper.

#### New `RoutingResult`

The new result from `swap_map_trial` is no longer several distinct but interlinking objects, but an encapsulated `RoutingResult` that borrows the same data as `RoutingProblem` borrows (the `SabreDAG`, the `DAGCircuit`, etc), and includes the chosen swaps inline.  It contains all the information necessary to rebuild a fully routed circuit, but does not do this until requested.

### Implementation changes

#### Remaining Python-space passes

`SabreLayout` and `SabreSwap` in Python space mostly just organise the Python-space inputs into the correct data types for Rust space, pass everything off, then reconstruct the Python-space objects that those passes are expected to store in the `PropertySet`.  They contain next to no business logic themselves.  The exception is that `SabreLayout` still has its mode for running with a custom routing pass, which is still all in Python space.

#### `Target`-native

Both layout and routing are now `Target`-native.  The Rust methods only accept a `Target`.  If Python space is supplied with a `CouplingMap`, we lift it into a dummy `Target` that implies the same connectivity constraints.  The public attributes of the passes that allowed access to coupling-map properties are retained, lazily computed on demand, for those that might still be accessing them.

#### Disjoint handling

The disjoint handling is now done entirely in Rust; `sabre_layout_and_routing` handles it internally.  This includes full routing of the DAG, even if it is split across multiple components; the previous `SabreLayout` implementation bailed out at this stage and left it to a later `SabreSwap` pass, but it's only a couple more lines of code to handle in Rust space now.

A follow-up commit may hugely simplify the disjoint handling such that the DAG never even needs to be split up.



### Old details and comments

<details><summary>Old PR message, now out of date</summary>
As of commit 815bc3d95f8c4aa768068f3530d8594ced98afbb (the original HEAD of this PR), `SabreSwap` works (I believe) completely correctly end-to-end.  `SabreLayout` is currently totally broken in Python space because I haven't finished rewriting the pass yet.  `StarPrerouting` is straight-up deleted because it reaches deep into Sabre internals, and I didn't want to have to deal with the pass just to have a working build - it'll be restored at some point.

#### New `SabreDAG`

The new `SabreDAG` is far more detached from the Python-space one, and needed (or perhaps just _wanted_) its control-flow handling rewritten so the ownership model both of `SabreDAG` and the output `RoutingResult` was far clearer when it came to control-flow blocks.  Previously, the ownership around the temporary `DAGCircuit`s of the inner blocks was quite awkward in the Rust/Python split, with the blocks being flattened in the result, but Python space holding onto the temporary `DAGCircuit`s backing the blocks, in order to do the rebuild later.  The new `SabreDAG` stores the temporary `DAGCircuit`s within itself, next to the `SabreDAG` that refers to that block.  This part of the change made the move to pure-Rust result and rebuild operations much more natural and easier.  The types of node in the graph are now called `Synchronize`, `TwoQ` and `ControlFlow`, where `Synchronize` is an operation that has no routing requirements, it only imposes data-synchronisation constraints (like a barrier, say).

An additional change to `SabreDAG` that probably _shouldn't_ be in this PR (and I might separate out later) is the coalescence of runs of operations onto the same arguments, or subsets, into the same node in the Sabre interaction graph, wherever possible.  For example, any 1q gate is compressed into its preceding 2q gate.  Similarly, any time we see a 2q gate followed by zero or more 1q gates on those two qubits, then another 2q gate on the same two qubits, the second 2q gate imposes no possible further routing constraints, so is always compressed into the prior node.  I mostly made this change as a performance optimisation for `update_route` and `populate_extended_set` (there are fewer nodes to explicitly visit), and because I wanted to avoid the `HashMap`-based `SwapMap` strategy of the Python-space result object, and instead have the swaps inline.  The cost, though, is that I didn't want the result object to have loads of empty `Vec`s in it - coalescing the nodes in the interaction graph has the knock-on effect of making it so _most_ of the individual nodes in the result will have initial swaps.

#### New `RoutingResult`

This replaces the tuple of a few `pyclass`es, in favour of single unified Rust handling.  The `RoutingResult` object stores a reference to the `SabreDAG` (and the `DAGCircuit`) it relates to, so it can rebuild itself, and so that it needn't store copies of the full routing order; just as the prior Python result object stored node indices in the Python DAG to facilitate the rebuild, this one stores node indices in the `SabreDAG`, so we get the full order of coalesced nodes without cloning it into the result object.

The new `RoutingResult` object can completely reconstruct the physical output when acting in `sabre_routing` mode, or (in theory - untested as yet) `sabre_layout` mode when there is no disjoint handling required.  `RoutingResult::rebuild` is effectively a full implementation of circuit embedding, layout application and routing rebuild.  Going further, `RoutingResult::rebuild_onto` _should_ (again, untested) work to manage rebuilding onto a larger `DAGCircuit` after routing on only a subset of the `Target` (such as happens in disjoint-target handling).

#### New `RoutingTarget`

The previous `(NeighborTable, Graph<(), ()>, ArrayView2<f64>)` form of the target specification was strongly influenced by what we could create and pass around from Python, but also had a fair amount of duplication in it.  Previously we had the problem that the Rustworkx Python-exposed `CouplingMap` couldn't be consumed by our Rust passes, so we were creating a `NeighborTable` from a list of edges, _rebuilding_ a `Graph<(), ()>` from that, and calculating a distance matrix entirely separately.

Now we're Rust-native, we only want to work with a `Target` as input.

In the new form: the roles fulfilled by `NeighborTable` and `Graph<(), ()>` are combined into a single new `Neighbors` structure, which then implements all the `petgraph::visit` traits using `PhysicalQubit` as the index type, so can be passed natively to `dijkstra` (for the release valve) and `token_swapper` (for control-flow layout reversion), and in turn, the result swap paths come out already in terms of `PhysicalQubit`.

Internally, `Neighbors` is a flat list `[n0_0, n0_1, n1_0, n1_1, n1_2, n2_0, ...]`, where `nX_Y` means the `Y`th neighbor of qubit `X`, and a separate list of partition points into that flat list.  Fundamentally, it is the same as the list-of-lists approach of the old `NeighborTable`, just flattened so we don't even need `SmallVec` and its "did this spill to the heap?" runtime check on each access for data locality; the data is simply always localised because it is flat and contiguous.  It is similar to the CSR-like structure used by `SparseObservable`, for example.  `Neighbors` then implements all the suitable petgraph visitor traits as if it were an undirected graph.  Actually, it's more like a guaranteed-symmetric directed graph (each edge is duplicated) for performance reasons in swap-candidate enumeration with how Sabre works, but that largely doesn't matter.

The distance matrix is now calculated directly from `Neighbors`.  The relevant method isn't in `rustworkx_core` yet (see Qiskit/rustworkx#1439), so this patch contains a (re-optimised) port of that method.
</details>

jakelishman commented:
As of (commit hash will change in future rebases)
```
3de0913cc8 Ensure ancillas are allocated in sorted order
```
the tests (other than the disabled disjoint ones) should all pass.  Following commits affect the randomisation, and will be split off into separate PRs.

jakelishman committed:
Add further creation and destructuring methods to `NLayout`

These are to allow more efficient Rust-space use of the object, allowing
reallocations to be avoided when doing things like ancilla expansion and
layout transformations at a scale larger than token swapping.

jakelishman committed:
Rewrite Sabre interfaces to be Rust native

This is a major rewrite of all the Sabre interfaces, though the core
routing logic is unchanged.  Note that the test suite is unaltered.

At least as far as the test suite covers, this change is RNG compatible
with its parent, and in several places, explicit code is added to
achieve this.  Given the neighbour-tracking structures are updated,
there can be some divergences, however, depending on how the order that
a coupling map was defined.

Changes to data structures
==========================

Changed `SabreDAG`
------------------

`SabreDAG` is now completely Rust-internal and not even exposed to
Python.  It's constructed directly from a `DAGCircuit`.  When
control-flow blocks are encountered, the temporary `DAGCircuit` backing
them is stored within the `SabreDAG` node itself, so there's no need for
all the `block_results` side-car handling.

The nodes in the graph explicitly store how the node affects routing:
they can be `TwoQ` (the actual routing target), `Synchronize` (a
non-routing object that has only data dependencies) or `ControlFlow`.
This simplifies the internal handling within the router, since now it
doesn't need to re-infer it.  It also lets us decay control-flow
operations to simple `Synchronize` points when we're in layout mode,
similar to how clearing the `block_nodes` field could implicitly do
before.

A follow-up commit (which necessarily changes the RNG compatibility)
makes stronger use of this new `SabreDAG` form to compress the
interaction graph during construction, reducing the amount of work done
by routing trials when advancing the state after a gate is routed.

New `RoutingProblem`
--------------------

A minor record-like struct that bundles up several related arguments.

New `RoutingTarget`
-------------------

The old Python-space code had a 3-tuple

    (NeighborTable, Graph<(), ()>, ArrayView2<f64>)

to represent the hardware chip the routing was taking place on.  The
first two elements stem from difficulties transferring the Rustworkx
Python-wrapped graph backing `CouplingMap` down to Rust space, and from
trying to faithfully port the existing implementation to Rust.

The `NeighborTable` and the `Graph<(), ()>` functionally serve the same
role; they both relate a node to its neighbors, with one presenting a
list-like interface for efficient iteration, and one presenting a
graph-like interface so graph algorithms like `dijsktra`,
`token_swapper` and `adjacency_matrix` could be called on it.

Now in pure Rust space, both are replaced by a single `Neighbors`
structure, which supplies both interfaces.  It implements all the
relevant `petgraph::visit` traits, so it can be directly passed to
`petgraph` and `rustworkx_core` graph algorithms.  Its data format is a
flat CSR-like sorted list of neighbors. It's logically the same as
`NeighborTable`, except all the `SmallVec`s are flattened into a single
contiguous `Vec` with partition points.

The distance matrix remains the same; we need super fast lookups of
scores because it's in the innermost loop of the swap-mapper.

New `RoutingResult`
-------------------

The new result from `swap_map_trial` is no longer several distinct but
interlinking objects, but an encapsulated `RoutingResult` that borrows
the same data as `RoutingProblem` borrows (the `SabreDAG`, the
`DAGCircuit`, etc), and includes the chosen swaps inline.  It contains
all the information necessary to rebuild a fully routed circuit, but
does not do this until requested.

Implementation changes
======================

Remaining Python-space passes
-----------------------------

`SabreLayout` and `SabreSwap` in Python space mostly just organise the
Python-space inputs into the correct data types for Rust space,
pass everything off, then reconstruct the Python-space objects that
those passes are expected to store in the `PropertySet`.  They contain
next to no business logic themselves.  The exception is that
`SabreLayout` still has its mode for running with a custom routing pass,
which is still all in Python space.

`Target`-native
---------------

Both layout and routing are now `Target`-native.  The Rust methods only
accept a `Target`.  If Python space is supplied with a `CouplingMap`, we
lift it into a dummy `Target` that implies the same connectivity
constraints.  The public attributes of the passes that allowed access to
coupling-map properties are retained, lazily computed on demand, for
those that might still be accessing them.

Disjoint handling
-----------------

The disjoint handling is now done entirely in Rust;
`sabre_layout_and_routing` handles it internally.  This includes full
routing of the DAG, even if it is split across multiple components; the
previous `SabreLayout` implementation bailed out at this stage and left
it to a later `SabreSwap` pass, but it's only a couple more lines of
code to handle in Rust space now.

A follow-up commit may hugely simplify the disjoint handling such that
the DAG never even needs to be split up.

jakelishman commented:
As noted in the now-updated PR comment: there's still some minor work to fix up the last couple of `todo!` stubs, though I think that mostly they're still `todo!` because I wasn't immediately certain how they should be handled, and we don't have test-suite coverage of them.

I also haven't run benchmarks on this any time recently.  From a now-old run, I know that one of the ASV utility-scale benchmarks shows a regression, but I tracked it down to a miniscule deviation in (I think) floating-point stuff 10k+ steps into the Sabre routing algorithm.  Functionally, the only reason it appears at all is because #14244 is not fixed by this PR, and the dodgy extended-set tracking causes the heuristic to be _completely_ borked for huge circuits.

I have a couple of PRs to follow this one that break RNG compatibility.  They build on stuff that might seem a bit superfluous in the current PR - the reason is because I wrote way too much at once, and then modified the patch series back into an RNG-compatible patch followed by improvements, so some of the extra structures appear to get added before they're properly taken advantage of, but it was too hard to write another completely new implementation to get rid of them.

jakelishman commented:
Local benchmarking suggests that while this has expected performance improvements in `sabre_routing`, there's some sort of regression in `sabre_layout` that needs more investigation.

jakelishman commented:
No longer "on hold" for performance: the trouble was that this commit forcibly disabled double parallelisation in the routing component of a single layout trial, and on machines where the single layer of parallelism didn't saturate the thread count, that was a regression.  The intent had been to reduce parallelism synchronisation overhead, but in practice, each routing run takes long enough, and Rayon is efficient enough under these circumstances, that it's irrelevant.

jakelishman commented:
Changed benchmarks, running on the 192-thread benchmarking server.

```text
| Change   | Before [22367fca]    | After [a102ff63]    |   Ratio | Benchmark (Parameter)                                                                                   |
|----------|----------------------|---------------------|---------|---------------------------------------------------------------------------------------------------------|
| -        | 2.78±0.04ms          | 2.49±0.06ms         |    0.9  | circuit_construction.CircuitConstructionBench.time_circuit_copy(14, 131072)                             |
| -        | 2.80±0.03ms          | 2.52±0.01ms         |    0.9  | circuit_construction.CircuitConstructionBench.time_circuit_copy(20, 131072)                             |
| -        | 2.80±0.03ms          | 2.49±0.06ms         |    0.89 | circuit_construction.CircuitConstructionBench.time_circuit_copy(5, 131072)                              |
| -        | 2.80±0.02ms          | 2.38±0.05ms         |    0.85 | circuit_construction.CircuitConstructionBench.time_circuit_copy(8, 131072)                              |
| -        | 76.1±1ms             | 29.4±0.9ms          |    0.39 | mapping_passes.PassBenchmarks.time_sabre_swap(14, 1024)                                                 |
| -        | 123±3ms              | 47.8±1ms            |    0.39 | mapping_passes.PassBenchmarks.time_sabre_swap(20, 1024)                                                 |
| -        | 37.1±9ms             | 8.33±1ms            |    0.22 | mapping_passes.PassBenchmarks.time_sabre_swap(5, 1024)                                                  |
| -        | 37.8±0.5μs           | 34.3±0.08μs         |    0.91 | passes.PassBenchmarks.time_remove_diagonal_gates_before_measurement(20, 1024)                           |
| -        | 17.4±0.09μs          | 15.8±0.1μs          |    0.91 | passes.PassBenchmarks.time_remove_diagonal_gates_before_measurement(5, 1024)                            |
| -        | 10.1±0.02s           | 5.61±0.01s          |    0.55 | qft.LargeQFTMappingTimeBench.time_sabre_swap(1081, 'decay')                                             |
| -        | 9.76±0.01s           | 5.22±0s             |    0.53 | qft.LargeQFTMappingTimeBench.time_sabre_swap(1081, 'lookahead')                                         |
| -        | 81.7±10ms            | 33.4±0.1ms          |    0.41 | qft.LargeQFTMappingTimeBench.time_sabre_swap(115, 'decay')                                              |
| -        | 102±0.9ms            | 32.0±0.06ms         |    0.31 | qft.LargeQFTMappingTimeBench.time_sabre_swap(115, 'lookahead')                                          |
| -        | 1.23±0.02s           | 570±3ms             |    0.46 | qft.LargeQFTMappingTimeBench.time_sabre_swap(409, 'decay')                                              |
| -        | 1.23±0.01s           | 548±7ms             |    0.44 | qft.LargeQFTMappingTimeBench.time_sabre_swap(409, 'lookahead')                                          |
| -        | 5.85±0.06ms          | 5.22±0.02ms         |    0.89 | qft.QftTranspileBench.time_ibmq_backend_transpile(1)                                                    |
| -        | 6.85±0.2ms           | 5.99±0.02ms         |    0.88 | qft.QftTranspileBench.time_ibmq_backend_transpile(2)                                                    |
| -        | 2.88±0.01s           | 2.56±0.01s          |    0.89 | quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(1081, 10, 'lookahead')                |
| -        | 20.7±8ms             | 9.54±0.08ms         |    0.46 | quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(115, 10, 'decay')                     |
| -        | 20.3±0.2ms           | 9.30±0.06ms         |    0.46 | quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(115, 10, 'lookahead')                 |
| -        | 182±10ms             | 78.8±1ms            |    0.43 | quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(115, 100, 'decay')                    |
| -        | 178±8ms              | 74.7±0.7ms          |    0.42 | quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(115, 100, 'lookahead')                |
| -        | 276±10ms             | 192±0.2ms           |    0.69 | quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(409, 10, 'decay')                     |
| -        | 404±10ms             | 303±0.6ms           |    0.75 | quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(409, 10, 'lookahead')                 |
| -        | 14.7±4ms             | 6.70±0.02ms         |    0.46 | quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(1, 'synthesis')                       |
| -        | 16.3±2ms             | 5.66±0.02ms         |    0.35 | quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(1, 'translator')                      |
| -        | 16.3±8ms             | 7.45±0.03ms         |    0.46 | quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(2, 'synthesis')                       |
| -        | 8.21±1ms             | 6.44±0.02ms         |    0.78 | quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(2, 'translator')                      |
| -        | 149±5ms              | 124±2ms             |    0.83 | quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(20, 'synthesis')                      |
| -        | 7.42±0.06ms          | 6.52±0.02ms         |    0.88 | queko.QUEKOTranspilerBench.time_transpile_bigd(0, None)                                                 |
| -        | 6.38±3ms             | 5.12±0.01ms         |    0.8  | queko.QUEKOTranspilerBench.time_transpile_bigd(1, None)                                                 |
| -        | 8.25±0.2ms           | 7.11±0.06ms         |    0.86 | queko.QUEKOTranspilerBench.time_transpile_bigd(2, None)                                                 |
| -        | 22.3±8ms             | 10.8±0.04ms         |    0.48 | queko.QUEKOTranspilerBench.time_transpile_bntf(1, None)                                                 |
| -        | 84.8±5ms             | 70.9±0.7ms          |    0.84 | queko.QUEKOTranspilerBench.time_transpile_bntf(2, 'sabre')                                              |
| -        | 19.8±0.5ms           | 17.0±0.05ms         |    0.86 | queko.QUEKOTranspilerBench.time_transpile_bntf(2, None)                                                 |
| -        | 32.7±8ms             | 29.6±0.08ms         |    0.91 | queko.QUEKOTranspilerBench.time_transpile_bntf(3, None)                                                 |
| -        | 109±10ms             | 75.6±8ms            |    0.69 | queko.QUEKOTranspilerBench.time_transpile_bss(0, 'sabre')                                               |
| -        | 74.6±10ms            | 44.9±0.5ms          |    0.6  | queko.QUEKOTranspilerBench.time_transpile_bss(0, None)                                                  |
| -        | 20.2±10ms            | 15.9±0.04ms         |    0.79 | queko.QUEKOTranspilerBench.time_transpile_bss(1, None)                                                  |
| -        | 110±4ms              | 80.3±5ms            |    0.73 | queko.QUEKOTranspilerBench.time_transpile_bss(2, 'sabre')                                               |
| -        | 55.2±1ms             | 28.4±0.09ms         |    0.52 | queko.QUEKOTranspilerBench.time_transpile_bss(2, None)                                                  |
| -        | 211±10ms             | 130±9ms             |    0.62 | queko.QUEKOTranspilerBench.time_transpile_bss(3, 'sabre')                                               |
| -        | 413                  | 323                 |    0.78 | queko.QUEKOTranspilerBench.track_depth_bss_optimal_depth_100(1, 'sabre')                                |
| -        | 293                  | 239                 |    0.82 | queko.QUEKOTranspilerBench.track_depth_bss_optimal_depth_100(3, 'sabre')                                |
| -        | 7.87±0.05ms          | 7.14±0.02ms         |    0.91 | random_circuit_hex.BenchRandomCircuitHex.time_ibmq_backend_transpile(4)                                 |
| -        | 9.49±0.07ms          | 6.66±0.05ms         |    0.7  | ripple_adder.RippleAdderTranspile.time_transpile_square_grid_ripple_adder(10, 0)                        |
| -        | 17.0±0.2ms           | 10.9±0.05ms         |    0.64 | ripple_adder.RippleAdderTranspile.time_transpile_square_grid_ripple_adder(20, 0)                        |
| -        | 105±10ms             | 44.4±5ms            |    0.42 | ripple_adder.RippleAdderTranspile.time_transpile_square_grid_ripple_adder(20, 1)                        |
| -        | 96.4±2ms             | 60.8±5ms            |    0.63 | ripple_adder.RippleAdderTranspile.time_transpile_square_grid_ripple_adder(20, 2)                        |
| -        | 86                   | 78                  |    0.91 | statepreparation.StatePreparationTranspileBench.track_cnot_counts_after_mapping_to_ibmq_16_melbourne(6) |
| -        | 187                  | 161                 |    0.86 | statepreparation.StatePreparationTranspileBench.track_cnot_counts_after_mapping_to_ibmq_16_melbourne(7) |
| -        | 394                  | 342                 |    0.87 | statepreparation.StatePreparationTranspileBench.track_cnot_counts_after_mapping_to_ibmq_16_melbourne(8) |
| -        | 19.7±9ms             | 10.00±0.1ms         |    0.51 | transpiler_benchmarks.TranspilerBenchSuite.time_compile_from_large_qasm                                 |
| -        | 7.19±0.1ms           | 5.73±0.02ms         |    0.8  | transpiler_benchmarks.TranspilerBenchSuite.time_cx_compile                                              |
| -        | 14.4±4ms             | 6.02±0.02ms         |    0.42 | transpiler_benchmarks.TranspilerBenchSuite.time_single_gate_compile                                     |
| -        | 62.9±0.2ms           | 54.2±0.2ms          |    0.86 | transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(0)                    |
| -        | 159±2ms              | 137±2ms             |    0.86 | transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(1)                    |
| -        | 209±4ms              | 175±2ms             |    0.84 | transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(2)                    |
| -        | 42.0±8ms             | 18.3±0.07ms         |    0.44 | transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(1)                           |
| -        | 30.1±7ms             | 10.9±0.04ms         |    0.36 | transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(2)                           |
| -        | 24.8±10ms            | 13.3±0.02ms         |    0.54 | transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(3)                           |
| -        | 160±4ms              | 119±2ms             |    0.75 | transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(3)                                |
| -        | 728                  | 598                 |    0.82 | transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_qv_14_x_14(3)                         |
| -        | 5.41±0.04ms          | 3.84±0.04ms         |    0.71 | transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(0)                     |
| -        | 8.26±0.05ms          | 5.53±0.05ms         |    0.67 | transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(0)                     |
| -        | 71.0±5ms             | 54.0±0.9ms          |    0.76 | transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(2)                     |
| -        | 8.58±0.09ms          | 5.66±0.03ms         |    0.66 | transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(0)                         |
| -        | 25.3±0.7ms           | 7.73±0.03ms         |    0.31 | transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(2)                         |
| -        | 9.94±0.05ms          | 8.58±0.02ms         |    0.86 | transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(3)                         |
| -        | 423±5ms              | 371±3ms             |    0.88 | utility_scale.UtilityScaleBenchmarks.time_qft('cx')                                                     |
| -        | 533±10ms             | 481±1ms             |    0.9  | utility_scale.UtilityScaleBenchmarks.time_qft('cz')                                                     |
| -        | 536±10ms             | 484±2ms             |    0.9  | utility_scale.UtilityScaleBenchmarks.time_qft('ecr')                                                    |
| Change   |   Before [22367fca]  |   After [a102ff63]  |   Ratio | Benchmark (Parameter)                                                                   |
|----------|----------------------|---------------------|---------|-----------------------------------------------------------------------------------------|
| +        |                   54 |                  68 |    1.26 | queko.QUEKOTranspilerBench.track_depth_bigd_optimal_depth_45(1, 'sabre')                |
| +        |                 1209 |                1625 |    1.34 | queko.QUEKOTranspilerBench.track_depth_bntf_optimal_depth_25(0, None)                   |
| +        |                  270 |                 352 |    1.3  | queko.QUEKOTranspilerBench.track_depth_bntf_optimal_depth_25(1, 'sabre')                |
| +        |                  682 |                 760 |    1.11 | queko.QUEKOTranspilerBench.track_depth_bss_optimal_depth_100(0, 'sabre')                |
| +        |                  250 |                 359 |    1.44 | queko.QUEKOTranspilerBench.track_depth_bss_optimal_depth_100(2, 'sabre')                |
| +        |                  660 |                 997 |    1.51 | ripple_adder.RippleAdderTranspile.track_depth_transpile_square_grid_ripple_adder(20, 3) |
| +        |                  819 |                 925 |    1.13 | transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_qv_14_x_14(1)         |
| +        |                 2649 |                2982 |    1.13 | utility_scale.UtilityScaleBenchmarks.track_qv_depth('cx')                               |
| +        |                 2649 |                2982 |    1.13 | utility_scale.UtilityScaleBenchmarks.track_qv_depth('cz')                               |
| +        |                 2649 |                2982 |    1.13 | utility_scale.UtilityScaleBenchmarks.track_qv_depth('ecr')                              |
```

mtreinish commented on crates/circuit/src/packed_instruction.rs:
```diff
@@ -702,6 +702,19 @@ pub struct PackedInstruction {
+            op: gate.into(),
+            qubits,
+            clbits: Default::default(),
+            params: None,
```
`StandardGate` can have params. Is there a reason you didn't include it here?

jakelishman replied:
Whoops - no reason, just tunnel vision. I only use it for swap, and clearly wasn't paying attention.

mtreinish commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -4527,6 +4527,36 @@ impl DAGCircuit {
+        )?;
+        out.remove_qubits((0..out.num_qubits()).map(Qubit::new))?;
+        out.qargs_interner = Interner::new();
+        out.add_qreg(QuantumRegister::new_owning("q", num_qubits as u32))?;
```
I see the` q` here, I guess there's no avoiding it.

jakelishman replied:
It's necessary for equality tests, at a minimum, and we do have passes that require this.

mtreinish commented on crates/transpiler/src/passes/sabre/dag.rs:
```diff
@@ -0,0 +1,217 @@
+/// This interaction representation retains only information about routing necessities; when
+/// possible, nodes in the input [DAGCircuit] are combined into a single node for routing purposes
+/// (for example, 1q gates are always folded into a preceding node and runs of 2q gates on the same
+/// qubits are combined).
```
Oh this is a nice QoL improvement.

jakelishman replied:
I think this comment is actually a lie in this PR - I "undid" that behaviour for the first PR and it comes in a follow up, I think. I just missed this comment.

It's actually more than QoL - it significantly simplifies the extended-set handling, and stops is from double-counting 2q runs.

jakelishman replied:
Removed in c79ff2e0f6 (though it'll be back later).

mtreinish commented on crates/transpiler/src/passes/sabre/dag.rs:
```diff
@@ -0,0 +1,217 @@
+            .topological_op_nodes()
+            .expect("infallible if DAG is in a valid state")
+        {
+            let NodeType::Operation(inst) = &dag[dag_node] else {
+                panic!("op nodes should always be of type `Operation`");
+            };
```
You didn't want to use `.unwrap_operation()` here?

jakelishman replied:
I want to get rid of that method and all this handling with #13696, but in the meantime, personally I slightly prefer the "expect" form.

mtreinish commented on crates/transpiler/src/passes/sabre/dag.rs:
```diff
@@ -0,0 +1,217 @@
+        let mut initial = Vec::<NodeIndex>::new();
+        let mut sabre = DiGraph::new();
+        // The `NodeIndex`es in here are into `sabre`.
+        let mut wire_pos: HashMap<Wire, NodeIndex> = HashMap::with_capacity(dag.width());
```
This is definitely simpler to work with, but I'm surprised you didn't want to go for 3 vecs for qubits, clbits, and vars using the inner u32 value as the index. Although that's probably micro-optimizing something that isn't a bottleneck at all, so a hashmap is fine.

jakelishman replied:
`Wire` is the truest representative of the data-flow dependency, really - Sabre doesn't care what the data dependency is, just that there is one.

jakelishman replied:
Oh, you mean the performance. Yeah, overall this step is completely irrelevant for that.

mtreinish commented on crates/transpiler/src/passes/sabre/distance.rs:
```diff
@@ -0,0 +1,74 @@
+//
+// The implementation was modified to be generic over `petgraph::visit` traits, throw away the
+// handling of `StableGraph`, minimise allocations in the `bfs_traversal` call, and use a bitmap
+// data structure to track "seenness" rather than `HashSet`.
```
Do you want to throw a comment here that we can remove this when https://github.com/Qiskit/rustworkx/pull/1439 merges and is released (it's WIP because I want to steal the fixedbit set usage optimization, although I can wait and let you push up an optimize PR instead of co-authoring that PR).

jakelishman replied:
Done in b9f4cf947

mtreinish commented on crates/transpiler/src/passes/sabre/layout.rs:
```diff
@@ -9,223 +9,340 @@
+) -> PyResult<(DAGCircuit, NLayout, NLayout)> {
+    // TODO: verify `partial_layouts` are all in bounds.
+    let Some(num_physical_qubits) = target.num_qubits else {
+        todo!();
```
Don't forget this and maybe add test coverage for the failing case?

jakelishman replied:
Done in d53fa36fc5.

mtreinish commented on crates/transpiler/src/passes/sabre/layout.rs:
```diff
@@ -9,223 +9,340 @@
+    let allow_parallel = getenv_use_multiple_threads();
+    let coupling = match target.coupling_graph() {
+        Ok(coupling) => coupling,
+        Err(TargetCouplingError::AllToAll) => todo!(),
```
I assume this todo is to just expand the circuit to target width and return?

jakelishman replied:
Yeah, though I guess we should have a test too, since we clearly don't at the moment.

jakelishman replied:
Done in e53e3a5457.

mtreinish commented on crates/transpiler/src/passes/sabre/layout.rs:
```diff
@@ -9,223 +9,340 @@
+        layout: &NLayout,
+        mut qubit_fn: impl FnMut(PhysicalQubit) -> PhysicalQubit,
+    ) -> NLayout {
+        assert!(layout.num_qubits() <= num_qubits as usize);
```
Should this have a real error code? But also do we need to check this in release mode?

jakelishman replied:
It represents a calling bug of an internal function rather than a recoverable error, so it's a `panic`, but I think the logic of the function doesn't absolutely guarantee that we _would_ panic if it's wrong; it'd just silently do something broken.  I'll swap it to `debug_assert` without trouble.

jakelishman replied:
Done in 8236d6d088.

mtreinish commented on crates/transpiler/src/passes/sabre/layout.rs:
```diff
@@ -9,223 +9,340 @@
+                    // non-deterministic order.  Sorting it is a canonicalisation step, but the
+                    // exact order doesn't matter.  Sorted order happens to cause us to be RNG
+                    // compatible with the prior Sabre disjoint handling.
+                    subset.sort_unstable();
```
No `par_sort_unstable`? :D

mtreinish reviewed (commented):
I haven't made my way fully through the PR yet, but here are my comments part way though since I need to step away for other things for a while.

jakelishman commented:
Since this is retargeted to 2.2 to avoid overloading me, I'll move the preceding 7 commits into the standalone PRs they generally should be, and only weren't because this was trying to merge in a block.

jakelishman committed:
Safely handle zero-block control-flow ops

Python-space Qiskit doesn't actually allow specifying any control-flow
op with zero blocks right now, but this removes a `todo!` and makes us
safe against that assumption changing in the future.

jakelishman committed:
Handle case of uninitialized `Target`

jakelishman committed:
Handle out-of-range checks for partial layouts

jakelishman committed:
Handle implicit all-to-all targets

This generally does not come up within the preset pass-managers, because
we can skip layout selection if there's all-to-all connectivity.  If
`SabreLayout` is called directly, however, it is still good to have a
valid output.

jakelishman commented:
This still contains cherry-picks of #14491 and #14587, but once those are merged, this PR should be fully ready for review.

jakelishman commented:
This is now fully rebased and ready for review.

mtreinish commented on crates/transpiler/src/passes/sabre/layout.rs:
```diff
@@ -9,223 +9,356 @@
                         max_iterations,
                         num_swap_trials,
-                        run_in_parallel,
+                        allow_parallel && num_swap_trials > 1,
```
Does this functionally make a difference? I wouldn't expect rayon's overhead on a single element iterator  to be a big deal.

jakelishman replied:
I think it's possible this was left over from the initial version, where I'd rewritten the variables to deliberately forbid double nested parallelism.  I then had to remove that change because it was a regression for machines large enough to not saturate the threads, and I probably just didn't notice that this was changed.  I can change it back if you prefer.

mtreinish replied:
It's fine it doesn't really hurt anything, it also is in the swap_trial code or somewhere else internally. (I remember reviewing it on Friday).

mtreinish commented on crates/transpiler/src/passes/sabre/layout.rs:
```diff
@@ -9,223 +9,356 @@
+                                } else {
+                                    // TODO: this handling sucks, but it's better than panicking
+                                    // later in Sabre routing when nothing makes any sense.
+                                    Err(PyValueError::new_err("physical qubit on wrong chip"))
```
Wouldn't this be a bug in the disjoint handling and strictly internal? Should we treat this is as a rust space bug and just panic? Or was there a case where this was real and user actionable/recoverable?

jakelishman replied:
I think you can get into this situation by passing `partial_layouts`?  I think that was what I was worried about, and why I didn't panic - I think I had reason to believe (perhaps falsely) that it was achievable by a user / custom transpiler pass.

mtreinish replied:
Ah yeah that's fair I guess we don't do any validation of what a custom pass puts on the property set. So it's possible that someone put a random qubit value in a partial layout. I guess maybe a bit more detail in the error message here makes it clearer. Something like: "The partial layout {partial:?} contains the physical qubit {p} which does not conform to the correct qubit in a partial layout".

jakelishman replied:
I think we have _some_ amount of handling that attempts to split up the partial layouts correctly, so it's possible that this would be a bug in Qiskit if it ever triggered.  I'm sufficiently unclear that I wrote a wordier error message saying "here's a problem, maybe it's our fault, maybe it's yours" in 310b5f617e, and if it triggers, we can track it down better?

mtreinish commented on crates/transpiler/src/passes/sabre/layout.rs:
```diff
@@ -9,223 +9,356 @@
+            // will be duplicates in the list, and there may still be un-set entries, but that
+            // doesn't matter, because we only access physical qubits that come up.
+            let mut sub_from_full = vec![PhysicalQubit::new(u32::MAX); num_physical_qubits];
+            for component in &components {
```
I wonder if adding a layer of nested parallelism here would be worth it. We don't have to do it in this PR. But there should be no data dependency between the different components here (you could adjust how `sub_from_full` is handled pretty easily I think). But we can investigate that in a follow up as this will already be way faster than doing it in Python.

mtreinish commented on crates/transpiler/src/passes/sabre/layout.rs:
```diff
@@ -9,223 +9,356 @@
+                    full_layout[virt.index()] = component.physical_qubits[sub_phys.index()];
+                }
+            }
+            let max_virt = VirtualQubit::new(u32::MAX);
+            let max_phys = PhysicalQubit::new(u32::MAX);
```
Not that it matters but I'm asking because I don't know sure, could we use a `const` here? We'd have to make `new` a const fn if we wanted to use it here, but I was just wondering.

jakelishman replied:
We could, but also, I'd be hugely surprised if the local variable didn't already get constant-folded into all its use sites.

mtreinish commented on crates/transpiler/src/passes/sabre/neighbors.rs:
```diff
@@ -0,0 +1,373 @@
+        self.partition.len() - 1
+    }
+
+    /// Are two qubits neighbours?
```
:us: 

```suggestion
    /// Are two qubits neighbors?
```

jakelishman replied:
Done in 7d73ed6056.

mtreinish commented on crates/transpiler/src/passes/sabre/route.rs:
```diff
@@ -11,68 +11,357 @@
+        let mut blocks = self.control_flow.iter();
+        for node in &self.sabre.initial {
+            let NodeType::Operation(inst) = &self.dag[*node] else {
+                panic!("sabre DAG should only contain op nodes");
```
This would drive me crazy if I ever encountered this.

```suggestion
                panic!("Sabre DAG should only contain op nodes");
```

jakelishman replied:
Done in 7d73ed6056.

mtreinish commented on crates/transpiler/src/passes/sabre/route.rs:
```diff
@@ -11,68 +11,357 @@
+            }
+            let new_inst = PackedInstruction {
+                qubits: dag.insert_qargs(&apply_scratch),
+                ..inst.clone()
```
I didn't realize you could do this this is a handy syntax.

mtreinish commented on crates/transpiler/src/passes/sabre/route.rs:
```diff
@@ -11,68 +11,357 @@
+            }
+            let index = self.sabre.dag[item.node].index;
+            let NodeType::Operation(inst) = &self.dag[index] else {
+                panic!("sabre DAG should only contain op nodes");
```
```suggestion
                panic!("Sabre DAG should only contain op nodes");
```

jakelishman replied:
Done in 7d73ed6056.

mtreinish commented on crates/transpiler/src/passes/sabre/route.rs:
```diff
@@ -11,68 +11,357 @@
+                        .map(|q| VirtualQubit(q.index() as u32))
+                        .collect::<HashSet<_>>();
+                    // TODO: we're assuming that `DAGCircuit::remove_qubits` retains relative
+                    // ordering of the remaining qubits, but the method doesn't commit to that.
```
Where are we using remove qubits here? The one at L236?

jakelishman replied:
Yeah - the logic of "tracking the qubits to retain" starts all the way up here, which is presumably why I wrote the comment there, but I can reword to make it clearer.

jakelishman replied:
Reworded in 7d73ed6056.

mtreinish commented on crates/transpiler/src/passes/sabre/route.rs:
```diff
@@ -11,68 +11,357 @@
+                            .collect::<Result<Vec<_>, _>>()?;
+
+                        let OperationRef::Instruction(py_inst) = inst.op.view() else {
+                            panic!("control-flow nodes must be PyInstruction");
```
```suggestion
                            panic!("Control-flow nodes must be PyInstruction");
```

jakelishman replied:
At least by Rust stdlib conventions, panic messages are lower-cased - e.g.

```text
[~/tmp/panic] (qiskit/3.13) (main #)
jake@ninetales$ cat src/main.rs
fn main() {
    vec![1, 2, 3][4];
}

[~/tmp/panic] (qiskit/3.13) (main #)
jake@ninetales$ cargo run --quiet

thread 'main' panicked at src/main.rs:2:18:
index out of bounds: the len is 3 but the index is 4
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

jakelishman replied:
(though I'll change the Sabre ones, since we always spell it like it's a proper noun)

mtreinish replied:
I mean it's actually an acronym `SABRE` but we almost never use it in all caps.

jakelishman replied:
Yeah, I refuse to use all-caps things for awkward constructed acronyms, but I use the capital letter Sabre as a nod to it being a sort of name haha

mtreinish commented on crates/transpiler/src/passes/sabre/route.rs:
```diff
@@ -98,136 +397,134 @@ impl RoutingState<'_, '_> {
                 }
             }
         }
+        assert!(
+            initial_swaps.is_none(),
+            "if initial swaps are given, at least one node must be known to be routable"
+        );
```
Wouldn't it be better to just do:
```rust
if to_visit.is_empty() && initial_swap.is_some() {
    panic!()
}
```
before the while loop.

jakelishman replied:
That would miss the case that `to_visit` included nodes, but none of them were routable - the swaps would be silently thrown out.

jakelishman replied:
(That particularly bug was actually the case I was most worried about making very noisy, because it'd cause a correctness hole.)

mtreinish commented on crates/transpiler/src/passes/sabre/route.rs:
```diff
@@ -441,132 +734,101 @@ impl RoutingState<'_, '_> {
 ///     control flow, and the fourth is a permutation, where `final_permution[i]` is the final
 ///     logical position of the qubit that began in position `i`.
```
This probably needs to be updated since it's documenting the old return format.

jakelishman replied:
Done in 7d73ed6056.

mtreinish commented on qiskit/transpiler/passes/layout/sabre_layout.py:
```diff
@@ -205,12 +200,16 @@ def run(self, dag):
+            )
 
-        # Choose a random initial_layout.
         if self.routing_pass is not None:
```
We probably should deprecate this option and code path. It was a neat idea back in the day but it doesn't seem super practical anymore.

jakelishman replied:
We can if you like (in a follow up), though I don't think it's been touched much at all, so it's probably not hurting us a huge amount to leave it there.

mtreinish replied:
I'm more worried about differing expectations in performance. For example, one thing I know has come up by those doing research around sabre is that the we don't support multiple trials if you specify a `routing_pass`. From a performance PoV it doesn't make sense but it's more just the experience of using the pass is sufficiently degraded with the `routing_pass` argument I think it's disingenuous to include it here. We could make it a standalone pass though, maybe we call the rust sabre `LightSabreLayout` and then make `SabreLayout` the version that operates in Python and can take a `routing_pass` argument?

jakelishman replied:
I don't think the current interface is great, but it's what we've had for a long time, and long since we moved the majority of Sabre to Rust space.  I feel like churning the interface now would just be unnecessary trouble for users.

mtreinish commented on qiskit/transpiler/passes/routing/sabre_swap.py:
```diff
@@ -233,233 +228,23 @@ def run(self, dag):
-            initial_layout,
-            self.trials,
-            self.seed,
+        initial_layout = NLayout.generate_trivial_layout(num_dag_qubits)
```
Did you want to include this in the wall time measurement?

jakelishman replied:
Moved in ea9402d31e.

mtreinish reviewed (commented):
Overall I think this LGTM, I don't fully trust my concentration going through it all as I feel like I lost some of the details on the internal data structure changes as I was going through it.
I left a few small inline questions and comments, mostly nits. But aside from those I think we're probably good to merge this.

jakelishman committed:
Update documentation

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>

jakelishman committed:
Move layout initialisation out of timing block

jakelishman commented:
> I feel like I lost some of the details on the internal data structure changes as I was going through it

Yeah, sorry about this - I think if I'd been more structured, it _was_ possible to do this in some slightly smaller chunks.  The trouble was that I didn't originally start this with the aim of making Sabre Rust-native, I think I started from wanting to make some of the changes to `SabreDAG` that come in later PRs, so I didn't chart a clean path through this at all.

jakelishman committed:
Make bail-out error message more descriptive

jakelishman committed:
Merge remote-tracking branch 'ibm/main' into sabre/no-py

jakelishman commented:
The logical merge conflict with #14589 should now be resolved.
